### DISCUSSION TITLE
Help me understand how `i18next-http-backend` works and why it sends request for each namespace

### DISCUSSION THREAD

**Discussion by yaman3bd** [i18next/next-i18next#2172] [Q&A] [answered]
Hey, 
I am using `next-18next` with `i18next-http-backend` and I am sending API calls to fetch the translations, but I found that the request is sent for each namespace in my config like if I have `["common", "titles"]` it would send a request for each group and when I change the page and I have three namespaces like: `["common", "test", "home"]` it will send request also for common it will not cache it, is this the expected behavior or is there something missing I have?

my config:
```js
const backend = require("i18next-http-backend/cjs");
const { axios } = require("./lib/axios");

module.exports = {
  debug: false,
  i18n: {
    locales: ["ar"],
    defaultLocale: "ar"
  },
  use: [backend],
  backend: {
    loadPath: "{{lng}}|{{ns}}",
    request: async (options, url, payload, callback) => {
      try {
        const [lng, ns] = url.split("|");
        //it will logged for each namespace
        console.log("ns", ns);
        await axios
          .get("/translations", {
            params: { group: ns }
          })
          .then((response) => {
            callback(null, {
              data: response.data.data[lng][ns],
              status: 200
            });
          })
          .catch((error) => {
            console.log(error.response);
          });
      } catch (e) {
        callback(null, {
          status: 500
        });
      }
    }
  },
  reloadOnPrerender: true,
  serializeConfig: false
};
```

my page:
```tsx
export const getServerSideProps: GetServerSideProps = async ({ locale }) => {
  return {
    props: {
      ...(await serverSideTranslations(locale ?? i18nextConfig.i18n.defaultLocale, ["common", "titles"], i18nextConfig))
    }
  };
};

export default function Test() {
  const { t } = useTranslation();

  return (
    <div>
      {t("titles:blog")}

      <Link href="/">{t("common:test")}</Link>
    </div>
  );
}
```

**Comment by adrai** [comment]
For caching have a look at:

https://github.com/i18next/i18next-http-backend/blob/master/example/next/next-i18next.config.js

https://www.i18next.com/how-to/caching#server-side-next.js-caching-with-filesystem

**Reply by yaman3bd** [reply]
@adrai, thanks for the examples and I apologize for this late replay, I looked through the examples you gave me none of them worked,
when I try cache using SSR filesystem:
```js
const ChainedBackend = require('i18next-chained-backend')
const FSBackend = require('i18next-fs-backend/cjs')
const HttpBackend = require('i18next-http-backend/cjs')

const {instance: axios} = require("./lib/axios");

module.exports = {
    debug: false,
    i18n: {
        locales: ["ar"],
        defaultLocale: "ar"
    },
    use: [ChainedBackend],
    ns: ["common"],
    backend: {
        backends: [
            FSBackend,
            HttpBackend
        ],
        backendOptions: [{
            expirationTime: 7 * 24 * 60 * 60 * 1000,
        }, {
            loadPath: "{{lng}}|{{ns}}",
            request: async (options, url, payload, callback) => {
                try {
                    const [lng, ns] = url.split("|");
                    await axios.get("/translations", {
                        params: {ns},
                    }).then((response) => {
                        callback(null, {
                            data: response.data.data[lng][ns],
                            status: 200
                        });
                    }).catch((error) => {
                        console.log(error.response);
                    });
                } catch (e) {
                    callback(null, {
                        status: 500
                    });
                }
            }
        }],
    },
    serializeConfig: false
};
```
it troughs this error:
```
./node_modules/i18next-fs-backend/cjs/fs.js:2:0
Module not found: Can't resolve 'fs'
Did you mean './fs'?
```
I looked into this [issue](https://github.com/i18next/next-i18next/issues/935) but unfortunately, none of the mentioned solutions worked for me.
here is how I set up my project:
in `_app.tsx`:
```ts
export default appWithTranslation(App, i18nextConfig)
```
in my other pages:
```ts
import {GetServerSideProps} from "next";
import {serverSideTranslations} from "next-i18next/serverSideTranslations";
import i18nextConfig from "@/next-i18next.config";
import {useTranslation} from "next-i18next";

export const getServerSideProps: GetServerSideProps = async ({locale}) => {
    const translations = (await serverSideTranslations(locale ?? i18nextConfig.i18n.defaultLocale, ["titles"], i18nextConfig))

    return {
        props: {
            ...translations
        }
    };
}

export default function Home() {
    const {t} = useTranslation();

    return (
        <main
            className="flex min-h-screen flex-col items-center justify-between p-24"
        >
            {t("titles:awesome")}
        </main>
    )
}


```

also if this approach only could be run on the server side how would I cache the translations on the client side?
I tried to use `i18next-localstorage-backend` with the same config in this [example](https://github.com/i18next/i18next-http-backend/blob/master/example/next/next-i18next.config.js)
but it did not work too so it gives me this error
```
Error: Default namespace not found at public/locales/ar/common.json
```
I want to load the translations from the database not from `public` dir

could you tell me what is wrong with the config/setup I have or if there is something else I am missing?

**Reply by yaman3bd** [reply]
I tried to use this config to see if I can cache the translations on both the server side and the client side with the same setup I have:
```js
const ChainedBackend = require('i18next-chained-backend').default
const FSBackend = require('i18next-fs-backend/cjs')
const HttpBackend = require('i18next-http-backend/cjs')
const LocalStorageBackend = require('i18next-localstorage-backend').default

const {instance: axios} = require("./lib/axios");
const isBrowser = typeof window !== 'undefined'

module.exports = {
    debug: false,
    i18n: {
        locales: ["ar"],
        defaultLocale: "ar"
    },
    use: [ChainedBackend],
    ns: ["common"],
    backend: {
        backends: isBrowser ? [LocalStorageBackend, HttpBackend] : [FSBackend, HttpBackend],
        backendOptions: [{
            expirationTime: 7 * 24 * 60 * 60 * 1000,
        }, {
            loadPath: "{{lng}}|{{ns}}",
            request: async (options, url, payload, callback) => {
                try {
                    const [lng, ns] = url.split("|");
                    await axios.get("/translations", {
                        params: {ns},
                    }).then((response) => {
                        callback(null, {
                            data: response.data.data[lng][ns],
                            status: 200
                        });
                    }).catch((error) => {
                        console.log(error.response);
                    });
                } catch (e) {
                    callback(null, {
                        status: 500
                    });
                }
            }
        }],
    },
    serializeConfig: false
};
```
it still throws this error `Module not found: Can't resolve 'fs'`
is it possible to cache the translations on both sides? or do I only have to choose one approach?

**Reply by adrai** [reply]
you need to check if the config is loaded on client side or on server side... on client side you can't use i18next-fs-backend... so check for typeof window for example and load the backend conditionally, like in the examples...

**Reply by yaman3bd** [reply]
I am doing this already:
```
backends: isBrowser ? [LocalStorageBackend, HttpBackend] : [FSBackend, HttpBackend],
```
but I am getting the errors I have already mentioned:
`Module not found: Can't resolve 'fs'` when I use ` i18next-fs-backend`
and I try to use only `i18next-localstorage-backend`:
`Error: Default namespace not found at public/locales/ar/common.json`
I want to load the translations from the database not from public dir

**Reply by adrai** [reply]
The require statement needs to be done after the isBrowser condition... directly in the uses array

**Reply by yaman3bd** [reply]
I have updated my config but still the same errors:
```js
const ChainedBackend = require('i18next-chained-backend')

const HttpBackend = require('i18next-http-backend/cjs')

const {instance: axios} = require("./lib/axios");


const isBrowser = typeof window !== 'undefined'
const isDev = process.env.NODE_ENV === 'development'

const backendOptions = {
    loadPath: "{{lng}}|{{ns}}",
    request: async (options, url, payload, callback) => {
        try {
            const [lng, ns] = url.split("|");
            await axios.get("/translations", {
                params: {ns},
            }).then((response) => {
                callback(null, {
                    data: response.data.data[lng][ns],
                    status: 200
                });
            }).catch((error) => {
                console.log(error.response);
            });
        } catch (e) {
            callback(null, {
                status: 500
            });
        }
    }
}

const backend = {
    backends: [HttpBackend],
    backendOptions: [backendOptions]
};

if (isBrowser) {
    const LocalStorageBackend = require('i18next-localstorage-backend').default
    backend.backends.unshift(LocalStorageBackend)
    backend.backendOptions.unshift({
        expirationTime: 60 * 60 * 1000,
        prefix: 'i18next_res_'
    })
} else {
    const FSBackend = require('i18next-fs-backend/cjs')
    backend.backends.unshift(FSBackend)
    backend.backendOptions.unshift({
        expirationTime: 60 * 60 * 1000,
        loadPath: './public/locales_cache/{{lng}}/{{ns}}.json',
    })
}

module.exports = {
    debug: isDev,
    i18n: {
        locales: ["ar"],
        defaultLocale: "ar"
    },
    use: [ChainedBackend],
    ns: ["common"],
    backend,
    serializeConfig: false
};
```
here is an example [repo](https://github.com/yaman3bd/local-test) could you please take a look and try to run the code and see what happens?

**Reply by adrai** [reply, marked as answer]
https://github.com/yaman3bd/local-test/pull/1/files

like here: https://github.com/locize/next-i18next-locize/tree/local-caching/next-i18next.config.js

**Reply by yaman3bd** [reply]
this is awesome, thanks really for helping me out achieving what I wanted.
but I have a few questions:
- I do not have the `locales_cache` dir in my project, and the keys are cached how does it work exactly?
- I have over 20 namespaces in my project and I do not need them all at once maybe the user won't visit a page that requires some namespaces so I did not include anything in `ns: ['namespaces']` so do I have to make sure to preload the namespaces or only the most used namespaces like for example: `common` or I could leave it empty and load the namespaces as required on each page?

**Reply by adrai** [reply]
On server side I recommend to always load everything at start. Once loaded it will use the resources in memory for all sub-sequent requests.

**Reply by yaman3bd** [reply]
but loading everything at the start won't affect the initial time load for the page?
also, the API request is called for every namespace individually so sending 20 requests to the API will block me because I have a rate limit, so I thought of a workaround to add namespace `ns: ['*']` so which means to load all the namespaces I have at the start and cache them, is there another way to do it?  how would you recommend handling such scenarios?

**Reply by adrai** [reply]
We are talking about next-i18next and the server side part here... so the best way for that is to preload all languages and namespaces... because on subsequent requests all languages are already there an don't need to be loaded again...
But naybe I misunderstood something. 🤷‍♂️

PS. also read this section: https://github.com/i18next/next-i18next#5-declaring-locale-dependencies

**Reply by yaman3bd** [reply]
I meant when I add all the namespaces I have, then a request will be sent for each namespace separately. if I have 20 ns then wait for the first request with the first ns to load and then add the response to the callback
then wait for the 2nd request and so on ...
so that means sending lots of requests one after another and this scenario has 2 possible cons:
my API  server will cause a rate limit error.
or maybe the internet connection in the client is slow so it will wait for all the requests to finish and this will cause a longer initial time to load the page.
so what I am thinking of is: preloading all the namespaces with one request by  adding `ns: ['*']` and cache each ns individually like:
```js
const ChainedBackend = require("i18next-chained-backend/dist/cjs/i18nextChainedBackend.js");
const FSBackend = require("i18next-fs-backend/cjs");
const HttpBackend = require("i18next-http-backend/cjs");

const { axios } = require("../lib/axios");

const NS = [
  "common",
  "home",
  "about",
  "contact",
  "footer",
  "header",
  "layout",
  "login",
  "register",
  "reset-password",
  "user"
];

module.exports = {
  i18n: {
    locales: ["ar"],
    defaultLocale: "ar"
  },
  use: [ChainedBackend],
  //it means send one request and preload all namespaces
  ns: ["*"],
  backend: {
    backends: [FSBackend, HttpBackend],
    backendOptions: [
      {
        expirationTime: 60 * 60 * 1000 // 1 hour
      },
      {
        loadPath: "{{lng}}|{{ns}}",
        request: async (options, url, payload, callback) => {
          try {
            const [lng, ns] = url.split("|");

            await axios
              .get("/translations")
              .then((response) => {
                //loop over all the namespaces and cache each namespace individually
                for (const [key, value] of Object.entries(response.data.data[lng])) {
                  if (NS.includes(key)) {
                    callback(null, {
                      data: response.data.data[lng][key],
                      status: 200
                    });
                  }
                }
              })
              .catch((error) => {
                // eslint-disable-next-line no-console
                console.log("response error: ", error.response.data);
              });
          } catch (e) {
            // eslint-disable-next-line no-console
            console.log("catch error: ", e);
            callback(null, {
              status: 500
            });
          }
        }
      }
    ]
  },
  serializeConfig: false
};
```
what do you think about this approach and how would you recommend handling this case?
please let me know if you need an example so I could update the example repo with the api-rate-limit and slow internet connection issues

and thanks really for helping me :pray:

**Reply by adrai** [reply]
Sounds very similar to: https://github.com/i18next/i18next-http-backend/issues/114 @bfgelectronics

**Reply by yaman3bd** [reply]
thanks for pointing it out, I tried to use `MultiloadAdapter` but it did not work maybe because I am using it with `ChainedBackend` and `FSBackend`?

here is [my config](https://github.com/yaman3bd/local-test/blob/main/next-i18next.config.js/node.js)
could you please take a look and tell me what I have wrong with my setup?

also in [this file](https://github.com/yaman3bd/local-test/blob/main/public/locales/ar/translatable.json) I have my translation object, I am loading it from another server but I added them to the public dir only for the testing so my question is that: is it ok to have everything in one file and then return only the required ns or do I have to make a separate file for each ns I have?

I really appreciate the effort you make for helping me out thanks a lot :pray:

**Reply by bfgelectronics** [reply]
You should really try my example (it has a small issue on the be side that I'll fix , but it works (meaning that it makes just one req per language))

**Reply by yaman3bd** [reply]
I tried it, but its not using cache layer,  in my project I am loading the translations from external API then use the cache layer `FSBackend` so I could not know how should I set everything up all together, could you update your example with the latest changes?
the example I provided it is exactly like my production project and setup
also, are you using any caching strategy in your project?

**Reply by yaman3bd** [reply]
just to be clear, here is how the translations work in my project:
I have the translations that the user can edit from the admin dashboard
and I have only one language `ar` so I do not have any problems with loading multi-language because I already have only one
but the issue I am having here loading more than one namespace at a time
does `MultiloadAdapter` supports loading all the namespaces for a given lang?
@bfgelectronics
